ユーザー管理機能

sisyphe0713 opened the pull request:
ユーザ管理機能の実装が完了しました。レビューお願いいたします。

#What
ユーザー管理機能(新規登録、ログイン、ログアウト)

#Why
ユーザーごとに個別にデータを保存するため

機能の様子: Gyazo GIF
https://gyazo.com/233011d7732c18aab135ba29dbe269b3

テストコードが通っている様子: Gyazo
https://gyazo.com/0b6480545fc241c5ded8870a7f776dee

sisyphe0713 committed:
Devise導入、配布された素材を導入

Deviseをインストール
Userモデル作成
マイグレーション
配布されたビューファイル、スタイルシート素材を導入

sisyphe0713 committed:
index画面表示(今やらなくてもよかった)

Itemモデルを作成
マイグレーション編集
マイグレーション実行
indexアクションのルーティング設定
itemsコントローラー作成
indexアクションをコントローラーに定義
ルートパスを設定

sisyphe0713 committed:
ビューファイルの導入のミスを解消

_second-header.html.erbを
誤ってviews/devise/sharedに入れていた。
これをviews/sharedに移動

ユーザー登録画面がlocalhost:3000/users/sign_inで表示されるようになった。

sisyphe0713 committed:
usersテーブルにカラム追加

views/registrations/new.html.erbでモデル名と新規登録機能へのパスを設定

usersテーブルにカラムを追加してマイグレート

views/registrations/new.html.erbのカラム名を設定

sisyphe0713 committed:
ストロングパラメーターの設定

sisyphe0713 committed:
画面遷移+バリデーション

-ログインユーザーのみにアクセスを許可
-ログインしていない場合のリダイレクト設定
-バリデーション設定
-stylesheetの配置を修正(reset.cssが反映されていなかった問題を解消)
-エラーが表示されるようビューを修正
-ログイン画面のビューの変数を修正
-ログインの有無でビューの表示を分岐

sisyphe0713 committed:
単体テストコード

-RSpec導入
-FactoryBot, faker導入
-authentication_user!いったん無効化
-バリデーション追加(ユーザー名全角、パスワード英数字混合)
-form_withにmodelとURLを記述→ログイン遷移できるように

sisyphe0713 committed:
空のcommit

kata-myu commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+    </h1>
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%= render 'shared/error_messages', model: f.object %>
+  <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
こちらのコメントアウトは不要かと思いますので、削除しておきましょう。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

kata-myu commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+<%= render "shared/second-header"%>
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
+<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
```
<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
こちらのコメントアウトは不要かと思いますので、削除しておきましょう。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

kata-myu commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  validates :nickname, presence: true
+  validates :last_name, presence: true, format: {with: /\A[ぁ-んァ-ン一-龥]/ }
+  validates :first_name, presence: true, format: {with: /\A[ぁ-んァ-ン一-龥]/ }
+  validates :last_name_kana, presence: true, format: {with: /\A[ぁ-んァ-ン一-龥]/ }
+  validates :first_name_kana, presence: true, format: {with: /\A[ぁ-んァ-ン一-龥]/ }
+  validates :birthday, presence: true
```
「presence: true」が重複しているため、こちらは「**with_options**」を使ってまとめてみましょう。

また、
**名前・名字のフリガナは全角カタカナ**となるように正規表現を修正しましょう。

修正をお願いいたします。

sisyphe0713 replied:
ご指摘ありがとうございます。
-with_optionsでまとめました。
-全角カタカナの正規表現に修正しました。

https://gyazo.com/07a60adda1392332f26470072dddb84a

sisyphe0713 replied:
ご指摘ありがとうございます。
-with_optionsでまとめました。
-正規表現修正しました。

kata-myu commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,121 @@
+require 'rails_helper'
```
こちらのファイルに以下のテストを追加してください。
```
①passwordとpassword_confirmationが一致していないと登録できない。
②名前のフリガナは、全角カタカナでないと登録できない。
③名字のフリガナは、全角カタカナでないと登録できない。
```

よろしくお願いいたします。

sisyphe0713 replied:
ご指摘ありがとうございます。
テストに追加し、動作確認しました。

https://gyazo.com/bed5bbe836f6f28e31704dce7e7d02d3

sisyphe0713 replied:
ご指摘ありがとうございます。
テストコード追加、修正しました。

kata-myu reviewed (commented):
レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認をお願いいたします。

sisyphe0713 committed:
レビューコメント対応

-不要なコメントアウト部(完了した指示)を削除
-with_optionsでバリデーションを簡潔化
-フリガナのバリデーションを全角から全角カタカナに修正
-テストコードを追加、修正

Saka-toku commented on app/models/user.rb:
```diff
@@ -0,0 +1,23 @@
+    end
+  end
+
+  # VALID_PASSWORD_REGEX =/\A(?=.*?[a-z])(?=.*?[A-Z])(?=.*?[\d])\w{6,12}\z/ #英小文字含む、#英大文字含む、数字含む、6文字以上
```
20行目のコメントアウトは削除しておきましょう！

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

Saka-toku commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,22 @@
+    first_name_kana {"ゼンカク"} #Fakerで全角の出力があるとよいのだが
+
+    birthday {Faker::Date.birthday}
+  end
```
こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめいたします。

sisyphe0713 replied:
ご指摘ありがとうございます。
このファイル内のコメントアウト、検討した結果全て削除することにしました。

Saka-toku commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,128 @@
+    before do
+      @user = FactoryBot.build(:user)
+    end
+
```
以下の記述のテストを追加しましょう！
```
・全ての値が正しければ登録できること
```
またテストコードは以下のようにcontextを使用して正常系と異常系とを分けて記述しましょう！ そうすることで、とても可読性の高いコードとなります。
  ```rb
  context 'ユーザー新規登録ができる時' do
    it '' do
    end
  end

  context 'ユーザー新規登録ができない時' do
    it '' do
    end
  end
```

sisyphe0713 replied:
ご指摘ありがとうございます。
追加しました。
正常系について、いまいちどこまでやるべきかがわからず、自分なりにここまで必要かと思うところまで追加しました。

Saka-toku replied:
テストを追加いただきありがとうございます！
ご自身で考えられテストを追加する。素晴らしいです！
記述いただいたテストは全て正しく、非常に確認しやすいのですが
これらのテストは以下の記述のみで確認することができます。
```rb
context 'ユーザー新規登録ができる時' do
  it "全ての値が正しければ登録できること" do
    expect(@user).to be_valid
  end
end

context 'ユーザー新規登録ができない時' do
```
私の指摘が分かりずらく申し訳ございませんでした。
可読性も上げることができますので、記述の変更をしていただければ幸いです。

Saka-toku commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,128 @@
+      @user.valid?
+      expect(@user.errors.full_messages).to include("Password is invalid")
+    end
+
```
より正確なテストを行うため
passwordに以下のテストを追加しましょう！
```
・全角では登録できないこと
（AAAA１１などでお願いいたします）
```

sisyphe0713 replied:
ご指摘ありがとうございます。
全角が混じるとダメなことを認識していませんでした。
バリデーションも修正して、テストコード追加しました。

Saka-toku reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします！
また以下の挙動確認もいたしますので、お手数ですが追加でご提出いただけますでしょうか？
```
・新規登録に成功する動画
・新規登録に失敗し、エラーが表示される動画
・ログインに成功する動画
・ログインに失敗し、エラーが表示される動画
・ログアウトに成功する動画
```

sisyphe0713 committed:
レビューコメント対応2

-不要なコメントアウトを削除
-テストコードに正常系を追加
-passwordに全角は許可されないことを反映(バリデーション修正、テストコード追加)

sisyphe0713 commented:
動作を表す動画を追加します。

・新規登録に成功する動画
https://gyazo.com/b04e54447dc4220cedebdb2c4dbf9860
・新規登録に失敗し、エラーが表示される動画
https://gyazo.com/0299aecd6d59fa27f2ca28e477cc4138
・ログインに成功する動画
https://gyazo.com/72fdb0e9be83e789033c9a40067e0cd3
・ログインに失敗し、エラーが表示される動画
https://gyazo.com/591a67586a2bf887c2b1f1baae2314f8
・ログアウトに成功する動画
https://gyazo.com/c7b818b1e3b7976130fcd73bd71b2484

Saka-toku reviewed (commented):
レビュー依頼ありがとうございます！
また丁寧にご返信いただきありがとうございます。修正箇所が非常に分かりやすいです。
一点、コメントいたしましたのでご確認をお願いいたします！

sisyphe0713 committed:
レビューコメント対応3

-テストコード正常系の簡潔化

sisyphe0713 commented:
ご指摘ありがとうございます。
修正しました。
正常系テストコードの意味が少しわかった気がします。どうもありがとうございました！

aoitech commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,9 @@
 class ApplicationController < ActionController::Base
+  # before_action :authenticate_user!
```
```ruby
# before_action :authenticate_user!
```
こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめします。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除します。

aoitech commented on app/controllers/items_controller.rb:
```diff
@@ -0,0 +1,13 @@
+  before_action :move_to_index, except: [:index, :show]
+
+  def index
+    @items = Item.all
```
```ruby
before_action :move_to_index, except: [:index, :show]
```


```ruby
 @items = Item.all
```
```ruby

def move_to_index
    unless user_signed_in?
      redirect_to action: :index
    end
end	
```
こちらの記述は、今回のユーザー管理機能では使わない記述かと存じます。
削除かコメントアウトしておきましょう。

理由：
今後の実装で差分が上がらなくなることを防ぐためでございます。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除します。
(スタイルシートの配置場所を間違えてビューがおかしい時に勘違いしていろいろさまよった痕跡です^^;)

aoitech commented on app/models/user.rb:
```diff
@@ -0,0 +1,22 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
```
```ruby
# Include default devise modules. Others available are:
# :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
```
こちらのコメントアウトも不要であれば削除しておきましょう。

sisyphe0713 replied:
ご指摘ありがとうございます。
削除しました。

aoitech reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加で3点コメントいたしました。非常に細かい部分で恐縮ではございますが、ご確認いただけますと幸いです。
お手数をお掛けいたしますが、よろしくお願いいたします。

sisyphe0713 committed:
レビューコメント対応4

-不要なコメントアウトを削除
-今回の実装に関係のない部分を削除

ghost commented:
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
このブランチをマージし、masterブランチでプルを行いましょう！
機能が反映されていることを確認し、引き続き実装頑張っていきましょう！

sisyphe0713 commented:
LGTMありがとうございます。

みなさまレビューどうもありがとうございました。
色々気づくことができて理解が深まり感謝しています！
